Fixes #31439 - Add new/edit page in React

ofedoren opened the pull request:
Requires to work properly: https://github.com/theforeman/foreman/pull/8181. Not sure what minimal version of Foreman should we lock this plugin on: 2.3 as it is now or 2.4. Depends on when exactly this feature will be available for plugins...

This PR adds New and Edit webhook pages. I didn't split it to two PRs because those pages are pretty similar (main/form component the same, changes are minimal). But if it will be difficult to review (as any js-related PR :D) I'll split it.

From React perspective there are just 3 new main components: `NewPage`, `EditPage`, `WebhookForm`.
From the old UI perspective there are deletions only.

ofedoren commented:
@MariaAga, can I ask you for any kind of review here, please? :rose:

lzap commented:
Is that "yes will do" or "yes it's okay no comments" kind of a thumbs-up, @MariaAga ? ;-)

MariaAga commented:
I updated Oleh in chat that I'll review it next week :)

MariaAga commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookForm.js:
```diff
@@ -0,0 +1,182 @@
+
+  return (
+    <ForemanForm
+      onSubmit={(values, actions) => handleSubmit(values, actions)}
```
Can be just  `onSubmit={handleSubmit}`

MariaAga commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksEditPage/WebhooksEditPage.js:
```diff
@@ -0,0 +1,62 @@
+    },
+    breadcrumbItems: [
+      { caption: __('Webhooks'), url: WEBHOOKS_PATH },
+      { caption: `${__('Edit')} ${name}`, url: '' },
```
If the url is empty it can be removed

MariaAga reviewed (changes requested):
It's great seeing more React forms in our app, thanks @ofedoren! :tada: 
I sometimes had an issue where I would create a new webhook and it won't show up in the table (but did show up after a refresh), not sure what can cause it as it happened randomly.
Left some comment inline:

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/index.js:
```diff
@@ -0,0 +1,22 @@
+export default compose(
+  connect(mapStateToProps, mapDispatchToProps),
+  callOnMount(({ fetchTemplates }) => fetchTemplates())
+)(WebhookForm);
```
I've removed Edit/New page entirely. There is new/edit modal instead.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksEditPage/index.js:
```diff
@@ -0,0 +1,60 @@
+    event: selectWebhookValues(state).event,
+    enabled: selectWebhookValues(state).enabled,
+    verify_ssl: selectWebhookValues(state).verify_ssl,
+    ssl_ca_certs: selectWebhookValues(state).ssl_ca_certs || undefined,
```
Done for getting the values. Initial values are still in snake case for formik to submit valid params.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/Components/ForemanFormikField.js:
```diff
@@ -0,0 +1,79 @@
+    {({
+      field, // { name, value, onChange, onBlur }
+      form: { touched, errors }, // also values, setXXXX, handleXXXX, dirty, isValid, status, etc.
+      meta,
```
Done.

ofedoren commented on config/routes.rb:
```diff
@@ -1,12 +1,13 @@
     end
   end
   match '/webhooks' => 'react#index', via: :get
+  match '/webhooks/new' => 'react#index', via: :get
```
Removed for new/edit since there is only one actual page.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookFormSelectors.js:
```diff
@@ -0,0 +1,33 @@
+export const selectHasError = (state, key) =>
+  selectWebhookFormStatus(state, key) === STATUS.ERROR;
+
+export const selectIsLoading = (state, key) => {
```
You're right :) refactored.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksEditPage/WebhooksEditPageSelectors.js:
```diff
@@ -0,0 +1,32 @@
+  selectAPIResponse(state, WEBHOOK_API_REQUEST_KEY) || {};
+
+export const selectWebhookTemplateId = state =>
+  toString((selectWebhookValues(state).webhook_template || {}).id || '');
```
Done, thanks.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksPageActions.js:
```diff
@@ -20,6 +23,14 @@ export const initializeWebhooks = () => dispatch => {
 };
 
+export const fetchWebhook = id => (dispatch, getState) =>
+  dispatch(
```
Removed the method, there is no need in it anymore.

ofedoren commented:
@MariaAga, thank you for the review :rose:

I've updated and removed New/Edit pages. Now there is Index page only with possibility to create/update a webhook via modals. I think it looks and feels better then default CRUD Rails pages. Although I'd like to get some feedback from @lzap, @timogoebel also.

Two main problems now are:
 - One test will fail and I don't know how to fix it :(
 - If you try to switch tabs in Create/Edit modals it will call `handle_submit` instead of just switching the tabs. Can you or someone from @theforeman/ui-ux team point me where should I look. My guess that there is something worng with Foreman form based on Formik if the children of this component are not simple Fields. But my knowledge of React/Formik is limited to dig deeper :(

xprazak2 commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/__tests__/WebhooksIndexPage.test.js:
```diff
@@ -1,9 +1,10 @@
 import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
-import WebhooksIndexPage from '../WebhooksIndexPage';
+import WebhooksIndexPage from '../../WebhooksIndexPage';
```
I think this change causes the following test failure:

![index-test](https://user-images.githubusercontent.com/2664090/108473422-4b595400-728e-11eb-98ab-ef4e075f61ed.png)

You are now importing `WebhooksIndexPage/index.js` which is a connected component and complains about missing store in tests. If you keep the original line, it will import `WebhooksIndexPage/WebhooksIndexPage.js` and that test should no longer crash.

ofedoren replied:
well... I thought so too, but after refactoring: `could not find react-redux context value; please ensure the component is wrapped in a <Provider>`

xprazak2 replied:
I can see it now, it should work when you move redux-related stuff into the parent component.

A couple of things to move around:
1. remove `const dispatch = useDispatch();`  out of `WebhooksIndexPage.js`
2. move `handleCreateSubmit` from `WebhooksdIndexPage.js` to `WebhooksIndexPage/index.js` and bind it to dispatch there, then pass it through props.
3. Call to `useForemanModal` will need to be moved to `index.js` as well, I think you can do something similar like `callWithToastsContext`.

Let me know if you need more details on this.

xprazak2 commented:
This is a nice work, preventing the default action when handling the tab click takes care of premature form submission for me:

```js
  const handleTabClick = (event, tabIndex) => {
    event.preventDefault();
    setActiveTab(tabIndex);
  };
```

I get a couple of errors in console when interacting with the form, for example when filling the name field, so you might want to check those.

Also note that there is currently a breakage in forms infrastructure and some changes might be needed once we get the [fix](https://github.com/theforeman/foreman/pull/8318) in.

ofedoren commented:
> I get a couple of errors in console when interacting with the form, for example when filling the name field, so you might want to check those.

Yeap, forgot to mention those. Not sure what to do with them either. There is also not pleasant behaviour for template `select`. I'll try to use `TypeAheadSelect`, but probably this fail for ForemanForm...

xprazak2 commented:
Fixes for errors:

1. complaints about `formProps` passed to `Tabs` when form modal opens:

![form-props](https://user-images.githubusercontent.com/2664090/108679380-bacc7f00-74ec-11eb-9eaf-532f8ade4edd.png)

 `ForemanForm` expects form fields as children and [passes that prop to each child](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/forms/ForemanForm/ForemanForm.js#L51), but the fields here are inside tab and `Tab` component does not expect to receive `formProps`.

You can fix it here by using a custom inner component:

```js
<ForemanForm
      onSubmit={handleSubmit}
      initialValues={initialValues}
      validationSchema={webhookFormSchema}
      onCancel={onCancel}
      enableReinitialize
    >
      <FormTabs
        activeTab={activeTab}
        handleTabClick={handleTabClick}
        webhookTemplates={webhookTemplates}
        httpMethods={httpMethods}
        availableEvents={availableEvents}
      />
    </ForemanForm>

//
const FormTabs = ({formProps, disabled, activeTab, handleTabClick, webhookTemplates, httpMethods, availableEvents}) => {
  return (
    <Tabs activeKey={activeTab} onSelect={handleTabClick}>
  // ...
    </Tabs>
}
```
or we need to make `ForemanForm` more configurable in core.

2. Complaints about controlled/uncontrolled component when changing name:
![uncontrolled-name](https://user-images.githubusercontent.com/2664090/108681594-9faf3e80-74ef-11eb-9bf0-2b093f62a274.png)

Having empty string as initial value takes care of this, same for target URL.

```js
// WebhookCreateModal.js
  const initialWebhookValues = {
    http_method: 'POST',
    enabled: true,
    verify_ssl: true,
    http_content_type: 'application/json',
    event: 'host_created.event.foreman',
    name: "",
   target_url: ""
  };


```

> There is also not pleasant behaviour for template `select`.

Right, I would avoid using the jQuery based select2 in React as it is bound to cause problems. You can try how well TypeAheadSelect works in forms or go with a simple select field.

MariaAga commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/Components/ForemanFormikField.js:
```diff
@@ -0,0 +1,78 @@
+ForemanFormikField.defaultProps = {
+  required: false,
+  labelHelp: null,
+  inputSizeClass: 'col-md-4',
```
As the form is inside a small modal maybe this should be changed to 8 or 10?
I used a short url and it wasn't fully visible 
![Screenshot from 2021-02-22 18-04-40](https://user-images.githubusercontent.com/30431079/108734760-723ab300-7538-11eb-9cb8-3dfe20f0f384.png)

ofedoren replied:
Totally makes sense, changed to `8`.

MariaAga reviewed (commented):
Nice idea using a form for it! I also had an issue with using select2
```
select2.js:2007 Uncaught TypeError: Cannot read property 'get' of null
    at constructor.opening (select2.js:2007)
    at constructor.open (select2.js:1375)
```
The toast error says `Events can't be blank` but there is no field named events so it's confusing.
Using 
```
  const handleTabClick = (event, tabIndex) => {
    event.preventDefault();
    setActiveTab(tabIndex);
  };
```
Solved the submit on tab click issue for me as well.

ofedoren commented on app/controllers/api/v2/webhooks_controller.rb:
```diff
@@ -57,6 +57,11 @@ def update
         process_response @webhook.destroy
       end
+
+      api :GET, '/webhooks/events', N_('List available events for subscription')
```
Added new API endpoint to get the available events. Mainly for UI, because in API/Hammer all available events should be shown in API docs/command line help.

ofedoren commented on app/controllers/concerns/foreman_webhooks/controller/parameters/webhook.rb:
```diff
@@ -26,6 +26,7 @@ def webhook_params_filter
         end
 
         def webhook_params
+          params['webhook']['event'] = params['event'] unless params['event'].blank?
```
This should fix the error `events can't be blank` even if it was selected in the form. Not sure why it doesn't get into permitted params.

ofedoren commented on app/services/foreman_webhooks/webhook_service.rb:
```diff
@@ -65,7 +65,7 @@ def self.request(url:, payload: '', http_method: :GET, user: nil, password: nil,
 
       request = Object.const_get("Net::HTTP::#{http_method.to_s.capitalize}").new(uri.request_uri)
-      request.basic_auth(user, password) if user && password
+      request.basic_auth(user, password) if !user.blank? && !password.blank?
```
This is because new UI throws errors to console if the form has such fields `undefined` or `null`. So they should be `""`. But after form submit a webhook is being created with user/password being set alothugh it should be `nil`. As a workaround I check if user/password is `blank`, so it behaives correctly with empty strings.

ofedoren commented on lib/foreman_webhooks/engine.rb:
```diff
@@ -15,7 +15,7 @@ class Engine < ::Rails::Engine
     initializer 'foreman_webhooks.register_plugin', before: :finisher_hook do |_app|
       Foreman::Plugin.register :foreman_webhooks do
-        requires_foreman '>= 2.3'
+        requires_foreman '>= 2.4'
```
@xprazak2, do you know by a chance if I should update it to `2.5` due to https://github.com/theforeman/foreman/pull/8318 or the fix will be CPed into 2.4/2.3?

xprazak2 replied:
The fix should definitely be CPed into 2.4.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/Components/ForemanFormikField.js:
```diff
@@ -0,0 +1,113 @@
+import PropTypes from 'prop-types';
+import { Field as FormikField } from 'formik';
+import { Spinner } from '@patternfly/react-core';
+import { TypeAheadSelect } from 'patternfly-react';
```
I don't use Foreman's `TypeAheadSelect` because it doesn't support re-initialization and it was much simplier to use the common one.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/Components/ForemanFormikField.js:
```diff
@@ -0,0 +1,113 @@
+      field, // { name, value, onChange, onBlur }
+      form: { touched, errors, setFieldValue, setFieldTouched }, // also values, setXXXX, handleXXXX, dirty, isValid, status, etc.
+    }) => {
+      const defaultSelection = (fieldValue, initialOptions) => {
```
This is for `TypeAheadSelect` where it accepts array of object while a Formik field's value should be a simple value. Mostly for a better UX.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookForm.js:
```diff
@@ -0,0 +1,73 @@
+
+const webhookFormSchema = Yup.object().shape({
+  name: Yup.string().required(__('is required')),
+  target_url: Yup.string().required(__('is required')),
```
I've removed `.url` validator because of: https://github.com/jquense/yup/issues/224. If someone know something like `URI.regexp` in javascript, I'll add it here. For know the URL is being validated on the server.

ofedoren replied:
Actually, it will be better to leave it as it is, because any regex could not allow ERB tags in URL (which is supported by this plugin, or at least will be). Server-side validation only is OK for now.

ofedoren reviewed (commented):
Thanks a lot for the help and reviews, @MariaAga and @xprazak2! This was really helpful :heart:

The PR is now requires https://github.com/theforeman/foreman/pull/8318, without it the from will just stuck on submit or if an error occured. Also I wanted to ask if you know how I fix the selection: 
![ScreenShot-1614030366850](https://user-images.githubusercontent.com/32508194/108835589-fb50f900-75cf-11eb-813e-a6796ba6028f.png). It just looks weird.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookForm.js:
```diff
@@ -0,0 +1,185 @@
+      enableReinitialize
+    >
+      <Tabs activeKey={activeTab} onSelect={handleTabClick}>
+        <Tab eventKey={0} title={<TabTitleText>General</TabTitleText>}>
```
Sure, done.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookForm.js:
```diff
@@ -0,0 +1,185 @@
+    'subnet_destroyed',
+    'build_entered',
+    'build_exited',
+    'status_changed',
```
Removed `availableEvents` in favor the API call to fetch those.

jeremylenz commented:
> I wanted to ask if you know how I fix the selection

@ofedoren It looks like the inputs are extending the area of the enclosing element, and it's drawing a border around everything.  Weird.  We did have a similar problem in Katello (it was just a regular rectangle, but still a weird black border) and we solved it by adding a CSS rule
```css
*:focus {
   outline: none;
 }
```
in our CSS overrides.

ofedoren commented:
> > I wanted to ask if you know how I fix the selection
> 
> @ofedoren It looks like the inputs are extending the area of the enclosing element, and it's drawing a border around everything. Weird. We did have a similar problem in Katello (it was just a regular rectangle, but still a weird black border) and we solved it by adding a CSS rule
> 
> ```css
> *:focus {
>    outline: none;
>  }
> ```
> 
> in our CSS overrides.

Thanks! I wanted to use something similar, but `*` will take every component, which I thought is not quite right.

MariaAga commented:
@ofedoren I think you can add the css rule to `pf-c-tab-content` and maybe add a custom classname to the tabs so it will choose the plugin rule instead of pf4s rule

ofedoren commented:
Everything seems to be addressed, so after https://github.com/theforeman/foreman/pull/8318 is merged, this can be merged also.

MariaAga commented:
@ofedoren After an error from the backend, the form remains open and loading so the user can't edit it and resubmit, I see here 2 solutions:
1. close the form on submit
2. Allow to edit after submit if the form is open

![Screenshot from 2021-03-02 13-02-16](https://user-images.githubusercontent.com/30431079/109639370-c6a4ea80-7b57-11eb-84cc-4f3fb5866299.png)

ofedoren commented:
> @ofedoren After an error from the backend, the form remains open and loading so the user can't edit it and resubmit, I see here 2 solutions:
> 
> 1. close the form on submit
> 2. Allow to edit after submit if the form is open
> 
> ![Screenshot from 2021-03-02 13-02-16](https://user-images.githubusercontent.com/30431079/109639370-c6a4ea80-7b57-11eb-84cc-4f3fb5866299.png)
> 
> Edit:
> looks like the form also doesn't close on success which it should :)

Hm.. I thought I've fixed that. Did you check the PR with the one in the core?

MariaAga commented:
@ofedoren This is also happening with the pr from core

MariaAga commented:
I took a look and it might be an issue with the core pr

MariaAga reviewed (changes requested):
Found the issue source + some comnets:
* auto choosing `subscribe to` makes it look like there aren't other options and I think that might confused users (I thought that I only have "Host Created" option)
* `Optional HTTP headers as JSON (ERB allowed)` value doesn't get saved for some reason.

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/Components/WebhookForm/WebhookForm.js:
```diff
@@ -0,0 +1,71 @@
+
+  return (
+    <ForemanForm
+      onSubmit={handleSubmit}
+      initialValues={initialValues}
+      validationSchema={webhookFormSchema}
+      onCancel={onCancel}
+      onSuccess={onSuccess}
+      enableReinitialize
+    >
```
Weird, AFAIR `item` wasn't required before and it make sense... I'll change that

ofedoren commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/Components/WebhooksTable/Components/NameToEditCell.js:
```diff
@@ -0,0 +1,41 @@
+      isInline
+      isDisabled
+      component="span"
+      onClick={jest.fn()}
```
Sorry, overlooked autochange

ofedoren commented:
@MariaAga, thank you :rose:, everything should be addressed now.

I've removed default event for `Subscribe to` field and fixed problem with `HTTP Headers` field (value was being saved, but the value didn't propagate to the form afterwards)

Also, although it worked before, after some changes in the core PR it doesn't work now :( I've check with `Personal Access Tokens` and `Bookmarks` and it didn't work there as well.

MariaAga commented:
I don't know if the issue is in this PR or the core PR but if I edit a webhook, submit and then click on "Create webhook" the modal opens and immediately closes

ofedoren commented:
@MariaAga, I'd say let's just wait until the PR in core is merged, then we can fully address any changes here if needed.

ofedoren commented:
@MariaAga, rebased for https://github.com/theforeman/foreman/pull/8318. Seems to be working again :rocket:

laviro commented:
Awesome work @ofedoren !
I would suggest using [PF4 Empty State](https://patternfly-react.surge.sh/components/empty-state)
when there are no webhooks listed, not showing the search bar and the button in the Empty state would be used to create a new one ( can use the same modal )

laviro commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/Components/WebhookCreateModal.js:
```diff
@@ -0,0 +1,61 @@
+  onCancel: PropTypes.func.isRequired,
+};
+
+export default WebhookCreateModal;
```
small issue with the modal, when accidentally clicking outside of it, all the data that the user entered is gone :(
is it possible to disable modal close after click outside? I think that the PF4 modal disables it.

![Webhooks](https://user-images.githubusercontent.com/26363699/111625808-f4479000-87f5-11eb-9aa4-eb27deb049a6.gif)

lzap commented:
Just a quick headsup - I need to do a quick patch that must go in before this PR because the plugin does not work on SSL correctly for Foreman 2.4. I will link my PR in here, then I will cut a new release for 2.4 and we can finally merge this. BTW the patch will add one new boolean flag to the webhook and we will need to add it here too - I will make sure to do minimum possible changes to prevent breaking changes but there will very likely be a git conflict - sorry about that!

It looks great, just one UX nitpick or idea - I think that "Subscribe to" should be the very first field on the first (General) tab. When I was creating a webhook today I realized that I actually don't know what the options are until I reach this field, so I think it should go in this order:

* Subscribe to
* Name
* Target URL
* Template (I think we can drop Webhook from the label)

Also HTTP Method can probably go to Additional tab, not sure about this one.

lzap commented:
Here is the patch: https://github.com/theforeman/foreman_webhooks/pull/32

ofedoren commented:
Thanks, @laviro and @lzap!

@laviro, I've updated the usage of empty state we have in core (now it's applied not only on the table, but on the whole page), also fixed the form to not disappear on misclick.

@lzap, I've updated the PR accordingly to the latest master and your suggestions about the form. Although I've left `HTTP Method` field on the first page since it's required and also prefiled, so I guess no harm if it stays there.

Also, since the form regression wasn't CPed into 2.4 this feature requires Foreman 2.5; updated JS deps to 8.4.1 as well.

ares commented:
I think this is a great improvement overall. From user testing I found few details.

1) Could the modal window be a but wider so the field labels are not that squezed and taking multiple rows? attaching screenshots below

2) the selection of the event causes the form to run the validation, however I haven't typed anything yet, so three fields are imidiatelly turned red. That is not overly inviting, could this just validate the event field or not trigger any validations?

3) the small i icons with the additional tooltip help seems to be jumping to the new lines (perhaps the same issue as 1)

4) given the even and templates selectors are exact list not accepting custom strings, why we don't use select2 instead? that has also search capability and is used everywhere else in the application?

![wh1](https://user-images.githubusercontent.com/109773/114064040-87299680-9899-11eb-9b58-a14b607c5731.png)
![wh2](https://user-images.githubusercontent.com/109773/114064046-885ac380-9899-11eb-859f-bec5959a99f9.png)
![wh3](https://user-images.githubusercontent.com/109773/114064048-885ac380-9899-11eb-807f-9af3e18798cf.png)

ofedoren commented:
> I think this is a great improvement overall. From user testing I found few details.
> 
> 1. Could the modal window be a but wider so the field labels are not that squezed and taking multiple rows? attaching screenshots below
> 2. the selection of the event causes the form to run the validation, however I haven't typed anything yet, so three fields are imidiatelly turned red. That is not overly inviting, could this just validate the event field or not trigger any validations?
> 3. the small i icons with the additional tooltip help seems to be jumping to the new lines (perhaps the same issue as 1)
> 4. given the even and templates selectors are exact list not accepting custom strings, why we don't use select2 instead? that has also search capability and is used everywhere else in the application?
> 

1. Agree, I need to revisit this. I hope Foreman modal's width can be adjusted.
2. I guess that's how Formik validation works. When the form was touched, run the validations. I'll see what I can do, but maybe @laviro can point me?
3. Got it.
4. Well, I thought that those selectors behave almost the same way, but don't use JQuery and look better (see https://github.com/theforeman/foreman_webhooks/pull/17#issuecomment-783214066). Also, I hate to say it, but we really lack standards or strict definitions for forms. There are some tips, guidelines, but there is no strict rule. Also, didn't we want to get rid of select2 to have something more PF styled?

jeremylenz commented:
> I hope Foreman modal's width can be adjusted

`ForemanModal` doesn't have any props etc. for adjusting the layout, but you should be able to do anything you need with CSS.  However, if you're using PF4 components here I'd recommend just using the modal component direct from `@patternfly/react-core`, since ForemanModal uses PF3 components anyway.

laviro reviewed (commented):
Great improvement!
as for the Form discussion, I think @ofedoren put a lot of effort into this implementation,
and I would be happy to see this work merged before the upcoming foreman code freeze...
I would suggest as a follow-up to explore how can we move the whole page into PF4, preferably by creating those component helpers in foreman core.
but for now, I think you can adjust some of the css so it would look wider and text won't collapse.

I think that after the last refactor something broke, and now if there's an error while submitting,
the user can't fix it.. and need to close and re-open the modal:
![Webhooks (1)](https://user-images.githubusercontent.com/26363699/114300620-e1a93980-9ac9-11eb-9024-9b5fccb39157.gif)

ofedoren commented:
@laviro, that's weird. I wouldn't submit PR changes if they break the main funcionality. It works without anything new:

![1](https://user-images.githubusercontent.com/32508194/114391060-51d1c100-9b97-11eb-9787-59aab40cf547.gif)

laviro reviewed (commented):
@ofedoren hey I did try again to submit->get error->fix it-> resubmit
and now it works good, probably some server-side issue at that time :D

as for the scss issue, you can pass the `className` prop to the `ForemanModal`,
and in the `scss` file you can do something like:

```scss
.webhooks-modal {
  .modal-dialog {
    width: 70vw;
    min-height: 70vw;
  }
}
```

ares commented:
> I guess that's how Formik validation works. When the form was touched, run the validations. I'll see what I can do, but maybe @laviro can point me?

I hope there's a way to improve, I didn't see such behavior in other react forms, but maybe there were no other required fields.

> Well, I thought that those selectors behave almost the same way, but don't use JQuery and look better (see #17 (comment)). Also, I hate to say it, but we really lack standards or strict definitions for forms. There are some tips, guidelines, but there is no strict rule. Also, didn't we want to get rid of select2 to have something more PF styled?

The main difference is the usability. I haven't tried myself yet, but what happens if I customize the string to non-existing event? Anyway not a blocker. If we're getting rid of select2, fine, but we should have an alternative. This field does not indicate it's a drop down, in other words a selector for picking one value from the predefined list.

ofedoren commented:
@ares, thanks for pushing on this. I've found where I had a bug causing this :facepalm:. Regarding select2, I think we can wait until we have proper tools in core so they can be used in plugins and we stick with the same patterns. As for now, the form cannot be submitted if wrong value was set.

Also, thanks @laviro for reminder about scss existence :D

laviro reviewed (commented):
All looks great! tested and seem to work as expected :)
My last suggestion would be to add an initial integration test, that we could upgrade over time,
just to see that we won't break stuff in the future, I know that it's a lot to ask for,
but an initial one would be good to have in this PR, here's a nice example: https://github.com/theforeman/foreman/pull/8419/files#diff-b66dddc7046163b1b7c1924a2289977b6538055911f8c7fffa4db8c54449c4a0

lzap commented:
Idea: Target URL could also suggest list of known smart proxy URLs, but it needs to allow entering an arbitrary URL too. Not sure if the component is capable of doing that tho, this would be just a nice to have.

Another detail: Headers render `\n` in the example, is it possible to render the example with real newlines? But if we are planning to write a new component for headers then it's probably fine.

![image](https://user-images.githubusercontent.com/49752/115225682-a7194f80-a10e-11eb-8b68-a20f017ec53f.png)

I wonder if we should also show the hook real name after the human-readable name, something like "Build Entered (foreman.hooks.host_managed.build)" because these are often shown in logs.

![image](https://user-images.githubusercontent.com/49752/115225280-32deac00-a10e-11eb-869f-7ba37ae61f26.png)

Looks great.

ofedoren commented:
@lzap, I agree that your suggestions are nice to have, but I'd vote to have them implemented in a separate PR(s). If you're OK with that, could you please open a RM issue(s)?

ofedoren commented:
@laviro, I've tried to add an integration test (see the last commit), but I failed to understand how to properly mock some functions/object. Currently the main problem is `history` in `Router` props (also it fails if I remove `Router` wrapper). Could you please point me? I've also tried to search for examples, but couldn't find any "whole page" examples.

laviro commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/__tests__/integration.test.js:
```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { IntegrationTestHelper } from '@theforeman/test';
+import { Router } from 'react-router-dom';
+
```
can you try to use `import { createMemoryHistory } from 'history'` ?

laviro commented on webpack/ForemanWebhooks/Routes/Webhooks/WebhooksIndexPage/__tests__/integration.test.js:
```diff
@@ -0,0 +1,26 @@
+  it('should flow', () => {
+    const integrationTestHelper = new IntegrationTestHelper();
+
+    const component = integrationTestHelper.mount(<Router history={{}}>
```
and then here you can use it in `history={createMemoryHistory()}`

laviro replied:
otherwise, you can pass here an object which contains all functionality you need, e.g:
```js
{
 listen: jest.fn,
 action: 'POP',
 etc....
}
```

laviro reviewed (commented):
@ofedoren oh, looks like we need to add here github action JS tests :D

laviro commented:
@ofedoren @lzap I just added JS test action in #34

ofedoren commented:
@laviro, so sorry to poke you again. I've somehow managed to pass that prop, but there is another issue now: `Error: Component(...): Nothing was returned from render.` Can you suggest what's the problem now?

laviro commented:
@ofedoren sorry for the delay, looks like there is a HOC mock issue with `withRenderHandler` (another reason to hate HOCs) :P
opened a [HOC mock fix](#36 ) PR to address that,
also, if you want we can meet and speed up this test debugging :)

ofedoren commented:
Thanks a lot again, @laviro! I must say it's a pain to mock all the needed components from core in the plugin...

Now it seems to be passing, but there is still a suppressed error which I don't know how to resolve..
`UnhandledPromiseRejectionWarning: Error: Warning: Cannot update a component (`%s`) while rendering a different component (`%s`).`

laviro commented:
> Thanks a lot again, @laviro! I must say it's a pain to mock all the needed components from core in the plugin...
> 
> Now it seems to be passing, but there is still a suppressed error which I don't know how to resolve..
> `UnhandledPromiseRejectionWarning: Error: Warning: Cannot update a component (`%s`) while rendering a different component (`%s`).`

It is indeed a pain, something we should explore more.. to use core JS code in plugins tests.

I feel like it's OK to merge it, and improve over time :)

laviro reviewed (approved):
Great work on this PR @ofedoren !!

lzap commented:
Are we good to merge then? I will give this a spin.

I am currently writing documentation, do not change anything from now on please :-)

lzap commented:
It works great, thank you all.
